providers/implementations/kem/hybrid_kem.c

```diff
@@ -0,0 +1,293 @@
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+const OSSL_PARAM *ossl_hybrid_kem_settable_ctx_params(void *vctx, void *provctx)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_settable_ctx_params(ctx, provctx);
+}
+
+const OSSL_PARAM *ossl_hybrid_kem_gettable_ctx_params(void *vctx, void *provctx)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_gettable_ctx_params(ctx, provctx);
+}
+
+int ossl_hybrid_kem_encapsulate(void *vctx, unsigned char *enc, size_t *enclen,
+                                unsigned char *secret, size_t *secretlen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t total_len = ossl_hybrid_get_key_length_from_ctx(ctx);
+    size_t a, b, encsum = 0, secretsum = 0;
+    size_t enclens[MAX_HYBRID_ALGS], seclens[MAX_HYBRID_ALGS];
+    unsigned int i;
+
+    if (total_len == 0)
+        return 0;
+    /* Calculate lengths */
+    for (i = 0; i < ctx->num_algs; i++) {
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], NULL, enclens + i,
+                                  NULL, seclens + i))
+            return 0;
+        encsum += enclens[i];
+        secretsum += seclens[i];
+    }
+    if (enc == NULL) {
+        if (enclen != NULL)
+            *enclen = encsum;
+        if (secretlen != NULL)
+            *secretlen = secretsum;
+        return enclen != NULL || secretlen != NULL;
+    }
+
+    /* Error checking */
+    if (*secretlen < secretsum) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*secretlen too small");
+        return 0;
+    }
+    if (*enclen < encsum) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*enclen too small");
+        return 0;
+    }
+
+    /* Set outputs */
+    *enclen = encsum;
+    *secretlen = secretsum;
+
+    /* Encapsulate */
+    for (i = 0; i < ctx->num_algs; i++) {
+        a = enclens[i];
+        b = seclens[i];
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], enc, &a, secret, &b))
+            return 0;
+        if (a != enclens[i] || b != seclens[i]) {
+            ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
+                           "secretlen or enclen changed");
+            return 0;
+        }
+        secret += seclens[i];
+        enc += enclens[i];
+    }
+    return 1;
+}
+
+int ossl_hybrid_kem_decapsulate(void *vctx,
+                                unsigned char *secret, size_t *secretlen,
+                                const unsigned char *enc, size_t enclen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t total_len = ossl_hybrid_get_key_length_from_ctx(ctx);
+    size_t a, encsum = 0, secretsum = 0;
+    size_t enclens[MAX_HYBRID_ALGS], seclens[MAX_HYBRID_ALGS];
+    unsigned int i;
+
+    if (total_len == 0)
+        return 0;
+    /* Calculate lengths */
+    for (i = 0; i < ctx->num_algs; i++) {
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], NULL, seclens + i,
```

Comment:
decapsulate? Or perhaps you really meant encapsulate here in order to calculate the enclens. In which case a comment would be really nice here. It's quite surprising to see decapsulate calling encapsulate

Reply:
My mistake, not that it matters.

Reply:
Comment added.